File: back/api/v1/endpoints/spents.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from crud import spent as crud_spent
from schemas.spent import SpentCreate, SpentRead, SpentUpdate
from api.deps import get_db
from datetime import date

router = APIRouter()

@router.post("/", response_model=SpentRead)
def create_spent(spent: SpentCreate, db: Session = Depends(get_db)):
    return crud_spent.create_spent(db=db, spent=spent)

@router.get("/", response_model=List[SpentRead])
def read_spents(
    user_id: int = Query(..., description="ID de l'utilisateur"),
    category_id: Optional[int] = Query(None),
    date_min: Optional[date] = Query(None),
    date_max: Optional[date] = Query(None),
    amount_min: Optional[float] = Query(None),
    amount_max: Optional[float] = Query(None),
    search: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    return crud_spent.get_spents(
        db=db,
        user_id=user_id,
        category_id=category_id,
        date_min=date_min,
        date_max=date_max,
        amount_min=amount_min,
        amount_max=amount_max,
        search=search
    )

@router.delete("/{spent_id}")
def delete_spent(spent_id: int, db: Session = Depends(get_db)):
    return crud_spent.delete_spent(spent_id=spent_id, db=db)

@router.put("/{spent_id}")
def update_spent(spent_id: int, spent_data: SpentUpdate, db: Session = Depends(get_db)):
    return crud_spent.update_spent(spent_id=spent_id, spent_data=spent_data, db=db)

@router.get("/total/")
def get_total_spent(year: int, user_id: int, month: Optional[int] = None, db: Session = Depends(get_db)):
    return crud_spent.get_total_spent(year=year, user_id=user_id, month=month, db=db)

@router.get("/total/income/")
def get_total_income(year: int, user_id: int, month: Optional[int] = None, db: Session = Depends(get_db)):
    return crud_spent.get_total_income(year=year, user_id=user_id, month=month, db=db)

@router.get("/total/remaining/")
def get_total_remaining(year: int, user_id: int, db: Session = Depends(get_db)):
    return crud_spent.get_total_remaining(year=year, user_id=user_id, db=db)

@router.get("/total/remaining/month")
def get_total_remaining_by_month(year: int, month: int, user_id: int, db: Session = Depends(get_db)):
    return crud_spent.get_total_remaining_by_month(year=year, user_id=user_id, month=month, db=db)

@router.get("/all/years/")
def get_all_years(user_id: int, db: Session = Depends(get_db)):
    return crud_spent.get_all_years(user_id=user_id, db=db)

@router.get("/total/by_category/")
def get_total_by_category(year: int, user_id: int, month: Optional[int] = None, db: Session = Depends(get_db)):
    return crud_spent.get_total_by_category(year=year, user_id=user_id, month=month, db=db)

@router.get("/year/income/")
def get_year_income(year: int, user_id: int, month: Optional[int] = None, db: Session = Depends(get_db)):
    return crud_spent.get_year_income(year=year, user_id=user_id, month=month, db=db)

@router.get("/year/spent/")
def get_year_spent(year: int, user_id: int, month: Optional[int] = None, db: Session = Depends(get_db)):
    return crud_spent.get_year_spent(year=year, user_id=user_id, month=month, db=db)

@router.get("/compare/month")
def compare_prev_month_spents(user_id: int, year: int, month: int, db: Session = Depends(get_db)):
    return crud_spent.compare_prev_month_spents(user_id=user_id, year=year, month=month, db=db)

@router.get("/compare/year")
def compare_prev_year_spents(user_id: int, year: int, db: Session = Depends(get_db)):
    return crud_spent.compare_prev_year_spents(user_id=user_id, year=year, db=db)

@router.get("/compare/income/month")
def compare_prev_month_income(user_id: int, year: int, month: int, db: Session = Depends(get_db)):
    return crud_spent.compare_prev_month_income(user_id=user_id, year=year, month=month, db=db)

@router.get("/compare/income/year")
def compare_prev_year_income(user_id: int, year: int, db: Session = Depends(get_db)):
    return crud_spent.compare_prev_year_income(user_id=user_id, year=year, db=db)

@router.get("/compare/remaining/month")
def compare_remaining_month(user_id: int, year: int, month: int, db: Session = Depends(get_db)):
    return crud_spent.compare_remaining_month(user_id=user_id, year=year, month=month, db=db)

@router.get("/compare/remaining/year")
def compare_remaining_year(user_id: int, year: int, db: Session = Depends(get_db)):
    return crud_spent.compare_remaining_year(user_id=user_id, year=year, db=db)

@router.get("/export")
def export_spents_csv(user_id: int, db: Session = Depends(get_db)):
    return crud_spent.export_spents_csv(user_id=user_id, db=db)

File: front/src/api/spents.js
import { UserProvider } from "../context/UserContext";
import instance from "./axios"

export const getSpents = async (id) => {
    const response = await instance.get(`/spents/?user_id=${id}`);
    return response.data
}

export const postSpents = async (payload) => {
    const response = await instance.post("/spents/", payload);
    return response.data;
}

export const deleteSpent = async (id) => {
    const response = await instance.delete(`/spents/${id}`)
    return response.data
}

export const modifySpent = async (id, payload) => {
    const response = await instance.put(`/spents/${id}`, payload)
    return response.data
}

export const totalSpent = async (id, year, month) => {
    let url = `/spents/total/?year=${year}&user_id=${id}`
    if (month) url += `&month=${month}`
    const response = await instance.get(url)
    return response.data
}

export const totalIncome = async (id, year, month) => {
    let url = `/spents/total/income/?year=${year}&user_id=${id}`
    if (month) url += `&month=${month}`
    const response = await instance.get(url)
    return response.data
}

export const totalRemaining = async (id, year, month) => {
    let url = `/spents/total/remaining/?year=${year}&user_id=${id}`
    if (month) url += `&month=${month}`
    const response = await instance.get(url)
    return response.data
}

export const totalRemainingByMonth = async (id, year, month) => {
    let url = `/spents/total/remaining/month/?year=${year}&month=${month}&user_id=${id}`
    const response = await instance.get(url)
    return response.data
}

export const allYears = async (id) => {
    const response = await instance.get(`/spents/all/years/?user_id=${id}`)
    return response.data
}

export const totalByCategory = async (id, year, month) => {
    let url = `/spents/total/by_category/?year=${year}&user_id=${id}`
    if (month) url += `&month=${month}`
    const response = await instance.get(url)
    return response.data
}

export const yearIncome = async (id, year) => {
    const response = await instance.get(`/spents/year/income/?year=${year}&user_id=${id}`)
    return response.data
}

export const yearSpent = async (id, year) => {
    const response = await instance.get(`/spents/year/spent/?year=${year}&user_id=${id}`)
    return response.data
}

export const compareMonthSpent = async (userId, year, month) => {
    const response = await instance.get(`/spents/compare/month?user_id=${userId}&year=${year}&month=${month}`)
    return response.data
}